File: Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param headA: the first list
     * @param headB: the second list
     * @return: a ListNode
     */
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        // write your code here
        ListNode dummyA(INT_MIN);
        ListNode dummyB(INT_MIN);
        dummyA.next = headA;
        dummyB.next = headB;
        headA = &dummyA;
        headB = &dummyB;
        while (headA && headB)
        {
            headA = headA->next;
            headB = headB->next;
        }
        ListNode *newHead = nullptr;
        if (headA)
        {
            newHead = &dummyA;
            while (headA)
            {
                headA = headA->next;
                newHead = newHead->next;
            }
            headA = newHead;
            headB = &dummyB;
        }
        else if(headB)
        {
            newHead = &dummyB;
            while (headB)
            {
                headB = headB->next;
                newHead = newHead->next;
            }
            headB = newHead;
            headA = &dummyA;
        }
        else
        {
            headA = &dummyA;
            headB = &dummyB;
        }
        
        while (headA && headB && headA != headB)
        {
            headA = headA->next;
            headB = headB->next;
        }
        
        if (headA == headB)
        {
            return headA;
        }
        else
        {
            return nullptr;
        }
    }
};

File: Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param headA: the first list
    # @param headB: the second list
    # @return: a ListNode
    def getIntersectionNode(self, headA, headB):
        # Write your code here
        
        pA, pB = headA, headB
        tailA, tailB = False, False
        
        while pA and pB:
            if pA == pB:
                return pA
            
            pA = pA.next
            if not pA and not tailA:
                pA = headB
                tailA = True
            
            pB = pB.next
            if not pB and not tailB:
                pB = headA
                tailB = True
        
        return None